### PR TITLE
Add checkin feature

### LINKED ISSUE #166: Check in feature
We discussed keeping the check-in system at the pub the other evening, as we don't think you should win a prize unless you actually attend!

I think it would be good for someone other than @asgrim to add this as a feature :)

### PR DESCRIPTION
**Pull request by asgrim**
Fixes #166 (sorry!)

### CODE REVIEW

File: src/App/Action/Account/Meetup/CancelCheckInAction.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        $this->entityManager = $entityManager;
+        $this->findMeetupByUuid = $findMeetupByUuid;
+        $this->findUserById = $findUserById;
+        $this->urlHelper = $urlHelper;
+    }
+
+    public function __invoke(Request $request, Response $response, callable $next = null) : Response
+    {
+        $meetup = $this->findMeetupByUuid->__invoke(Uuid::fromString($request->getAttribute('meetup')));
+        $user = $this->findUserById->__invoke(Uuid::fromString($request->getAttribute('user')));
```
Comment (lboynton) [inline src/App/Action/Account/Meetup/CancelCheckInAction.php]: The naming of `findUserById` seems a bit inconsistent with `findMeetupByUuid`? They are both searching by UUID right?

File: src/App/Entity/Meetup.php [hunk 2]
```diff
@@ -177,26 +176,64 @@ public function isBefore(\DateTimeImmutable $date) : bool
 
     public function attend(User $user) : void
     {
-        $this->attendees->add($user);
-        $user->meetupsAttended()->add($this);
+        foreach ($this->meetupAttendees as $meetupAttendee) {
+            /** @var MeetupAttendee $meetupAttendee */
+            if ($meetupAttendee->attendee()->getEmail() === $user->getEmail()) {
```
Comment (lboynton) [inline src/App/Entity/Meetup.php]: Is it possible to compare UUIDs to find the matching user?

File: src/App/Entity/MeetupAttendee.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ *     @ORM\UniqueConstraint(name="meetup_user", columns={"meetup_id", "user_id"})
+ * })
+ */
+/*final */class MeetupAttendee
+{
+    /**
+     * @ORM\Id
+     * @ORM\Column(name="id", type="guid")
+     * @ORM\GeneratedValue(strategy="NONE")
+     * @var string
```
Comment (lboynton) [inline src/App/Entity/MeetupAttendee.php]: `@var int`

File: src/App/Service/User/Exception/UserNotFound.php [hunk 4]
```diff
@@ -12,6 +13,11 @@ public static function fromEmail(string $email) : self
         return new self(sprintf('User with email "%s" was not found', $email));
     }
 
+    public static function fromId(UuidInterface $uuid) : self
+    {
+        return new self(sprintf('User with ID "%s" was not found', (string)$uuid));
```
Comment (lboynton) [inline src/App/Service/User/Exception/UserNotFound.php]: Is the cast here necessary?

Comment (asgrim) [inline src/App/Service/User/Exception/UserNotFound.php]: Yes, `$uuid` is an object and `sprintf` doesn't do casting I suppose

Comment (asgrim) [inline src/App/Service/User/Exception/UserNotFound.php]: https://3v4l.org/5Knq9

Comment (lboynton) [inline src/App/Service/User/Exception/UserNotFound.php]: Oh, news to me!

Comment (asgrim) [inline src/App/Service/User/Exception/UserNotFound.php]: I take it back, `__toString` makes it possible: https://3v4l.org/Pm82B
but still, no harm in it :)

Comment (lboynton) [inline src/App/Service/User/Exception/UserNotFound.php]: Ah, there's no `__toString()` by default?

Comment (asgrim) [inline src/App/Service/User/Exception/UserNotFound.php]: not by default, but `Ramsey\Uuid\Uuid` does actually implement it

**Comment by jezemery** [comment]
Looks good! I'll try and add the ajax part soon!

**Pull request merged**